File: elixir/word-count/lib/word_count.ex
defmodule WordCount do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t()) :: map
  def count(sentence) do
    add_string_to_acc = fn string, map ->
      key = String.downcase(string)
      Map.update(map, key, 1, &(&1 + 1))
    end

    sentence
    |> String.split(~r/[^A-Za-zÀ-ÖØ-öø-ÿ0-9\-]/, trim: true)
    |> Enum.reduce(%{}, add_string_to_acc)
  end
end

File: typescript/word-count/word-count.ts
type WordCount = Map<string, number>;

const newWordCount = () => { return new Map<string, number>(); };

export const phraseWords = (phrase: string): Array<string> => {
    return phrase
        .split(new RegExp("[^a-zA-Z0-9\u00c0-\u00d6\u00d8-\u00f6\u00f8-\uab6f]+"))
        .filter((e) => e != "");
}
const reducer = (counts: WordCount, word: string): WordCount => {
    let currentWord = word.toLowerCase();
    let currentCount: unknown = counts.get(currentWord);
    if (counts.has(currentWord))
        counts = counts.set(currentWord, (currentCount as number) + 1)
    else
        counts = counts.set(currentWord, 1);
    return counts;
}

export default class Words {
    count(s: string): WordCount {
        let words: Array<string> = phraseWords(s);
        return words.reduce(reducer, newWordCount());
    }
}

